content/learn/07.built-in-libraries/02.i2s/i2s.md

```diff
@@ -150,26 +165,67 @@ I2S.availableForWrite()
 none
 
 #### Returns
-the number of bytes available to write.
+size_t
+- the number of bytes available to write
 
 ### `onTransmit(handler)`
 
 #### Description
-Registers a function to be called when a block of data has been transmitted.
+Registers a function to be called when a block of data has been transmitted. This is used in conjunction with the two-argument function write(buf, len), and the callback function is called when the DMA write operation that was triggered by write(buf, len) finishes. The callback takes no arguments, which mean it cannot determine by itself if the write was completely successful or wrote only some part of the data. Partial writes need to be handled separately by the application.
+
+The callback is not triggered by the single argument version `write(val)`, only by the two argument version `write(buf, len)`. The callback can be registered once at setup of the I2S interface, and will be called for every subsequent write() -- you do not need to register it anew for every write operation.
 
 #### Parameters
-handler: the function to be called when data is sent and return nothing, e.g.: void myHandler()
+handler: the function to be called when data is sent. It must be a function with no arguments returning nothing, e.g.: void myHandler()
 
 #### Returns
 None
 
+### `read()`
+
+#### Description
+Reads binary data from the I2S interface. Data is read either as one sample or a sequence of samples.
+
+The single argument function `read(val)` reads one sample for one channel. A full read requires
+calling this function twice, once for the left channel and once for the right channel.
+If the device buffer is empty, the function blocks until there is data available.
+The return value is always an `int`. You will need to know the number of bits per sample
+to handle the data correctly.
+
+The two-argument function `read(buf, len)` reads a sequence of samples into the array `buf`.
+Data will be stored in the binary format for the current sample size, with the left and right
+channels interleaved. The argument `len` specifies the number of **bytes** in the buffer, not the
+number of samples. If the device has more data available than what fits in `buf`,
+the function retrieves as much as will fit. If the device has less data available, possibly
+nothing, the function will read what is available and return without blocking. The return
+value is the number of bytes that was actually read. It is the responsibility of the
+application to read the return value and act accordingly.
+
+#### Syntax
+
+```
+val = I2S.read()  // Might block until data is available for reading
+I2S.read(buf, len) // Will not block, but might not read all the data
```

Comment:
```suggestion
int val = I2S.read();  // Might block until data is available for reading
int bufferVal = I2S.read(buf, len); // Will not block, but might not read all the data
```

Reply:
Yes, of course. Sorry. I'm unfamiliar with the format for this documentation.